### PR TITLE
Changes to CBMC required to make gnat2goto work

### PR DESCRIPTION
**Pull request by NlightNFotis**
Hi.

These are changes that we had to make with Hannes in lockstep with gnat2goto while we were working on gnat2goto's tests and getting them to work. Without these, gnat2goto does produce output, but CBMC chokes on it.

### CODE REVIEW

File: src/util/Makefile [hunk 1]
```diff
@@ -39,6 +39,7 @@ SRC = arith_tools.cpp \
       json_irep.cpp \
       json_stream.cpp \
       json_symbol_table.cpp \
+      json_symbol.cpp \
```
Comment (hannes-steffenhagen-diffblue) [inline src/util/Makefile:42]: This is just fixing a makefile and probably should be in a separate commit

**Comment by hannes-steffenhagen-diffblue** [comment]
These changes are basically all "we apparently don't expect symbol types in a bunch of places where they could be popping up". We're not completely sure if that's something we need to fix in cbmc or the gnat2goto compiler, but doing it cbmc is "easier" for now (and isn't less correct than the previous version).

Though perhaps we should consider having a pass over the goto functions that resolves all symbol types once instead of doing separately in lots of places.

File: src/analyses/goto_check.cpp [hunk 2]
```diff
@@ -1439,24 +1439,26 @@ void goto_checkt::check_rec(const exprt &expr, guardt &guard, bool address)
...
-      if(!op.is_boolean())
+      auto op_followed = op;
+      op_followed.type() = ns.follow(op_followed.type());
+      if(op_followed.type().id() != ID_bool)
         throw "`"+expr.id_string()+"' takes Boolean operands only, but got "+
-              op.pretty();
+              op_followed.pretty();
 
-      check_rec(op, guard, false);
+      check_rec(op_followed, guard, false);
```
Comment (smowton) [inline src/analyses/goto_check.cpp:1456]: Why keep the resolved type rather than basing subexpressions on `op`? If you're trying to exclude symbol types that resolve to primitives because cbmc is fragile in the face of those then you should probably do that in a specific pass upon loading the symbol table, rather than doing it ad-hoc like this. Otherwise if you're happy with the symbol types but `is_boolean` isn't clever enough then only use the resolved type where you need it.

**Review by smowton — COMMENTED** [review]
Also suggest amending the PR title to actually describe what you've done

**Comment by smowton** [comment]
Re: symbol types btw, suggest consulting with Peter and/or Daniel. CBMC doesn't expect symbols that resolve to primitives apparently, which makes sense since that won't happen in C.
